library/alloc/tests/vec.rs

```diff
@@ -1095,6 +1094,18 @@ fn test_from_iter_specialization_panic_during_drop_leaks() {
     }
 }
 
+// regression test for issue #85322. Peekable previously implemented InPlaceIterable,
+// but due to an interaction with IntoIter's current Clone implementation it failed to uphold
+// the contract.
+#[test]
+fn test_collect_after_iterator_clone() {
+    let v = vec![0; 5];
+    let mut i = v.into_iter().peekable();
+    i.peek();
+    let v = i.clone().collect::<Vec<_>>();
+    assert!(v.len() <= v.capacity());
+}
+
```

Comment:
I was wondering if this test could be improved. Right now this detects the problem only because `IntoIter::clone` allocates a vec with the exact capacity needed but if it allocated extra capacity **at the end** of the vec this test would actually pass even though the `Peekable` implementation is still wrong. A more robust test should also check that it doesn't overwrite each element before reading it. For example:

```suggestion
#[test]
fn test_collect_after_iterator_clone() {
    let v = vec![0; 5];
    let mut i = v.into_iter().map(|i| i + 1).peekable();
    i.peek();
    let v = i.clone().collect::<Vec<_>>();
    assert_eq!(v, [1, 1, 1, 1, 1]);
    assert!(v.len() <= v.capacity());
}
```

With the current implementation the first assert fails due to `v` being equal to `[1, 2, 3, 4, 5]`. Another option would be changing the `0` to something like `String::from("abc")` and hope that the test catches the quintuple free.